### ISSUE TITLE
Incorrect opening_hours correction suggestion

### ISSUE THREAD

**Issue by Phyks** [osm-fr/osmose-backend#555]
Hi,

See https://osmose.openstreetmap.fr/en/map/#item=3250&zoom=17&lat=48.834572&lon=2.351347&level=1%2C2%2C3&tags=&fixable= , there is an `opening_hours` error reported on https://www.openstreetmap.org/way/161119956. The current `opening_hours` value is

> Tu-Su 11:00-18:00; Dec 25 off; Jan 01 off; May 01 off; Dec 31 11:00-17:00

It however seems to be valid according to https://projets.pavie.info/yohours/?oh=Tu-Su%2011:00-18:00;%20Dec%2025%20off;%20Jan%2001%20off;%20May%2001%20off;%20Dec%2031%2011:00-17:00. I did not dig very much in the details of this syntax to find out whether this is a false positive or not, yet.

The main issue is that the proposed fix is:

> opening_hours = Tu-Su 11:00-18:00; Dec 01:00 off; Jan 01:00 off; May 01:00 off; Dec 31 11:00-17:00 

which transforms "25" and "01" as days into hours ("01:00"), which is an error.

Thanks!
Best,

**Comment by frodrigo** [comment]
@HubMiner we re-enable the opening_hours on the web frontend, yesterday. Unfortunately there is already new issue.

**Comment by HubMiner** [comment]
Hi frodrigo,
this is the main false positive I have been noticing as well.  You can assign this to me.

BTW, is there a way to see OH false positives submitted to Osmose?

**Comment by frodrigo** [comment]
Does it require we disable this item again, or not ?

You can get the false positive here: http://osmose.openstreetmap.fr/en/errors/false-positive?item=3250
But not the full details.

**Comment by HubMiner** [comment]
Disabling: please give me few days to see how involved this is.

Revisiting false positives:
Ideally this list should represent issues that need to be fixed in Osmose.  Is there a way to reevaluate the false positives  all together with the current version of checker, or I would need to manually deleted  each false positive issue and it will then show up on the map again if the checker is not happy?

Do the false positives have to be handled separately by each region or is there a way to group them across regions?

Merci!

**Comment by frodrigo** [comment]
> Is there a way to reevaluate the false positives all together with the current version of checker

No

> or I would need to manually deleted each false positive issue and it will then show up on the map again

Please, don't do this.

> Do the false positives have to be handled separately by each region or is there a way to group them across regions?

Yes, you can use the filters: http://osmose.openstreetmap.fr/en/errors/false-positive?item=3250&country=switzerland

You can export the list as CSV and from the object id to test in one way or an other: http://osmose.openstreetmap.fr/en/errors/false-positive.csv?item=3250

**Comment by Phyks** [comment]
> Does it require we disable this item again, or not ?

From my (limited) experience so far, I fixed quite a bunch of incorrect opening hours syntax and only found this particular false positive. So, I would say this analysis is still very useful and worth keeping (maybe with a warning?) although there is this minor catch.

**Comment by HubMiner** [comment]
I made progress debugging this, see "oh-debug" branch in my copy of the repo:
https://github.com/HubMiner/osmose-backend/tree/oh-debug
but nothing I tried worked, so I need help.

run: modules/oh_sanitizer/main.py
it will run just the failing test case, and I added debug code so you can see how lark grammar interprets failing case and passing case.  I assume the issue is either in wrong priority for the month-date rule, or there is a missing statement for multiple "Month Date State, Month Date State" combination.

**Comment by frodrigo** [comment]
cc @rezemika

**Comment by rezemika** [comment]
It seems that priorities can only be assigned to terminals (and not to rules). So, `monthday_date.10` should not work (see https://lark-parser.readthedocs.io/en/latest/grammar/#terminals). What do you think of using a big terminal in `monthday_date_day` which would look like `MONTH " " INT` and to process it in a method? So, it could be possibe to assign a priority to it...

**Comment by HubMiner** [comment]
Thanks for great suggestion @rezemika, still no luck:

I created a separate "MONTH_DATE" TERMINAL with higher priority, but it looks like "hour_minutes" rule is still grabbing the value. This change now breaks previously working case, so both values are parsed incorrectly.

Do you have any hints on enabling debug in lark?

same repo: git diff  795b4650..3aa66fbc

**Comment by rezemika** [comment]
Oh, weird. A simple solution would be to make `hour_minutes` less permissive (make `:` mandatory, for example), but it would also break working cases...

Debug mode can be enabled by importing `logging` and using the `debug=True` parameter at the parser creation : https://lark-parser.readthedocs.io/en/latest/how_to_use/#lalr-usage

Also, I see that `hms` is still using priority ([here](https://github.com/osm-fr/osmose-backend/commit/3aa66fbc6059013008cfe1c741b9c3e214a3f3bd#diff-ad2fdfe380c4a216fee1c7a3ec40e6dcR142)), and it catches only two digits. It wouldn't be perfect, but make "AM" or "PM" mandatory may do the trick...

**Comment by fansanelli** [comment]
Hello,
I'm not sure it is the same bug, but I have another example:

Tu-Su,PH 15:30-18:30; Mo off; Jan 01 off; Dec 25 off

This is reported as invalid.
Can you please check this too?
Francesco

**Comment by rezemika** [comment]
I think the problem is that `PH` should be before `Tu-Sh` (see the specs [here](https://wiki.openstreetmap.org/wiki/Key:opening_hours/specification#weekday_selector)). ;)

**Comment by fansanelli** [comment]
Thanks @rezemika. I applied the suggested change. I hope will be fixed tomorrow :)

**Comment by HubMiner** [comment]
Hi @rezemika, I am not able to make any progress here, any chance you can take it?

**Comment by fansanelli** [comment]
@rezemika now osmose suggested this:

PH,Tu-Su 15:30-18:30; Mo off; Jan 01:00 off; Dec 25 off

Please note: Jan 01:00, but Jan 01 is the correct and currently used value.
This is a bug

**Comment by HubMiner** [comment]
This is consistent with the original bug, where date is interpreted as hours.

**Comment by HubMiner** [comment]
By the way, optional colon after date makes Osmose happy, but I don't recommend doing that globally.
Jan 1: off

**Comment by fansanelli** [comment]
I'll whitelist the osmose issue for now. Since I don't want to do a change not needed

**Comment by rezemika** [comment]
Oh weird, when I try so sanitize `Tu-Su,PH 15:30-18:30; Mo off; Jan 01 off; Dec 25 off` with OHSanitizer 0.1.9, I get `Tu-Su, PH 15:30-18:30; Mo off; Jan 1 off; Dec 25 off`. There's a problem with the coma, but the date is kept valid. (And it's the same thing with 0.1.10.)

Could you print the tree you get [here](https://github.com/osm-fr/osmose-backend/blob/master/modules/oh_sanitizer/main.py#L384) for problematic fields please? You can use `print(tree.pretty())` to get an indented result.

For `Tu-Su,PH 15:30-18:30; Mo off; Jan 01 off; Dec 25 off`, I get the following (with 0.1.9).

```
Tree(time_domain, [Tree(rule_sequence, [Tree(selector_sequence, [Tree(range_selectors, [Tree(weekday_or_holiday_sequence_selector, [Tree(weekday_sequence, [Tree(weekday_range, [Tree(wday, [Token(WDAY_TU, 'Tu')]), Tree(wday, [Token(WDAY_SU, 'Su')])])])])])])]), Token(RULE_SEPARATOR, ','), Tree(rule_sequence, [Tree(selector_sequence, [Tree(range_selectors, [Tree(weekday_or_holiday_sequence_selector, [Tree(holiday, [Token(PUBLIC_HOLIDAY, 'PH')])])]), Tree(time_selector, [Tree(timespan, [Tree(time, [Tree(hour_minutes, [Token(INT, '15'), Token(INT, '30')])]), Tree(time, [Tree(hour_minutes, [Token(INT, '18'), Token(INT, '30')])])])])])]), Token(RULE_SEPARATOR, ';'), Tree(rule_sequence, [Tree(selector_sequence, [Tree(range_selectors, [Tree(weekday_or_holiday_sequence_selector, [Tree(weekday_sequence, [Tree(weekday_range, [Tree(wday, [Token(WDAY_MO, 'Mo')])])])])])]), Tree(rule_modifier_off, [Token(OFF, 'off')])]), Token(RULE_SEPARATOR, ';'), Tree(rule_sequence, [Tree(selector_sequence, [Tree(range_selectors, [Tree(monthday_selector, [Tree(monthday_range, [Tree(monthday_date_monthday, [Token(MONTH, 'Jan'), Token(INT, '01')])])])])]), Tree(rule_modifier_off, [Token(OFF, 'off')])]), Token(RULE_SEPARATOR, ';'), Tree(rule_sequence, [Tree(selector_sequence, [Tree(range_selectors, [Tree(monthday_selector, [Tree(monthday_range, [Tree(monthday_date_monthday, [Token(MONTH, 'Dec'), Token(INT, '25')])])])])]), Tree(rule_modifier_off, [Token(OFF, 'off')])])])

time_domain
  rule_sequence
    selector_sequence
      range_selectors
        weekday_or_holiday_sequence_selector
          weekday_sequence
            weekday_range
              wday	Tu
              wday	Su
  ,
  rule_sequence
    selector_sequence
      range_selectors
        weekday_or_holiday_sequence_selector
          holiday	PH
      time_selector
        timespan
          time
            hour_minutes
              15
              30
          time
            hour_minutes
              18
              30
  ;
  rule_sequence
    selector_sequence
      range_selectors
        weekday_or_holiday_sequence_selector
          weekday_sequence
            weekday_range
              wday	Mo
    rule_modifier_off	off
  ;
  rule_sequence
    selector_sequence
      range_selectors
        monthday_selector
          monthday_range
            monthday_date_monthday
              Jan
              01
    rule_modifier_off	off
  ;
  rule_sequence
    selector_sequence
      range_selectors
        monthday_selector
          monthday_range
            monthday_date_monthday
              Dec
              25
    rule_modifier_off	off
```

**Comment by fansanelli** [comment]
Hello, I downloaded the file from master and did some test.
First I had to fix this issue:

https://github.com/lark-parser/lark/issues/383

And btw I still have this:

AssertionError: '24:00,07:00' != '24/7'

**Comment by frodrigo** [comment]
@fansanelli we are stuck in 0.6 https://github.com/osm-fr/osmose-backend/blob/master/requirements.txt#L8

**Comment by dfaure** [comment]
@fansanelli how did you fix that lark issue? I'm hitting the same in https://github.com/rezemika/humanized_opening_hours/issues/34

**Comment by frodrigo** [comment]
> @fansanelli how did you fix that lark issue? I'm hitting the same in [rezemika/humanized_opening_hours#34](https://github.com/rezemika/humanized_opening_hours/issues/34)

Use a 0.6 lark version. The grammar should be adjusted to use newer version. But a better option may be to use an other parser.

**Comment by rezemika** [comment]
I'm so sorry, I have almost no free time for a long time, and insufficient knowledge, to maintain this project clean and up to date. I would like to refactor it, but I lack of experience to write a good parser for such a complicated specification...

**Comment by dfaure** [comment]
So there's no good implementation in python of opening_hours? issue 925 generates tons of warnings (https://osmose.openstreetmap.fr/fr/byuser/davidfaure_bot?username=davidfaure_bot&item=3250), and fixing that is related to what's being said here: how to write a good parser in python. Would it be feasible to write one from scratch by porting https://github.com/opening-hours/opening_hours.js from Javascript to python? Or is that too dependent on a mountain of javascript-specific things?

**Comment by frodrigo** [comment]
> So there's no good implementation in python of opening_hours? issue 925 generates tons of warnings (https://osmose.openstreetmap.fr/fr/byuser/davidfaure_bot?username=davidfaure_bot&item=3250), and fixing that is related to what's being said here: how to write a good parser in python. Would it be feasible to write one from scratch by porting https://github.com/opening-hours/opening_hours.js from Javascript to python? Or is that too dependent on a mountain of javascript-specific things?

As I know the only full and reliable implementation is in Javascript.

But the current implementation in Osmose is not just a standard parser. It also support frequent mistake and so be able to point the error and suggest a fix (or not). The Osmose parser does not have to have to a full coverage, it can ignore part.

**Comment by dfaure-kdab** [comment]
Just to be 100% sure... we're discussing this here but the problems are in https://github.com/rezemika/oh_sanitizer right?

If so, I'm tempted to rewrite oh_sanitizer on top of the lex/yacc grammar that a colleague of mine just developed for a C++/Qt implementation, https://invent.kde.org/libraries/kopeninghours. This would require using a lex/yacc parser in python, like https://github.com/dabeaz/ply (does anyone have experience with that?). And then porting the "suggestion for fixes" logic. Maybe I should start with that last part in kopeninghours first to verify that the grammar is flexible enough for that.

Anyhow, does this sound like a good plan?
If so, should this happen as a mega MR for oh_sanitizer, or as a separate github repo that osmose can then switch to?

Disclaimer: I'm not very experienced with python, help or at least code review would be very welcome...

**Comment by frodrigo** [comment]
> Just to be 100% sure... we're discussing this here but the problems are in https://github.com/rezemika/oh_sanitizer right?

No, this repository wes abandoned, we fork and fix it in
https://github.com/osm-fr/osmose-backend/tree/master/modules/oh_sanitizer

> If so, I'm tempted to rewrite oh_sanitizer on top of the lex/yacc grammar that a colleague of mine just developed for a C++/Qt implementation, https://invent.kde.org/libraries/kopeninghours. This would require using a lex/yacc parser in python, like https://github.com/dabeaz/ply (does anyone have experience with that?). And then porting the "suggestion for fixes" logic. Maybe I should start with that last part in kopeninghours first to verify that the grammar is flexible enough for that.

I have experience with lex/yacc, but not in python. Feel free to use the tool you want.
Note, we already have an other parser for MapCSS, it is done with antlr.
https://github.com/osm-fr/osmose-backend/tree/master/mapcss

> Anyhow, does this sound like a good plan?
> If so, should this happen as a mega MR for oh_sanitizer, or as a separate github repo that osmose can then switch to?

As you wish.

> Disclaimer: I'm not very experienced with python, help or at least code review would be very welcome...

Sure, if you interested in the subject we will help you.

**Comment by dfaure-kdab** [comment]
Before I embark on a massive fork and language migration.... here's another thought: would it be possible for the osmose-backend script to run a C++ executable on the osmose server? That binary would simply take a string as input and output the sanitized string, and python would just have to do a `subprocess.check_output()` as far as I can see.
This would allow to truly reuse the existing implementation rather than forking it and having to maintain both in parallel.
For security reasons I guess I can't provide the binary myself, it would have to be built on the server (or by the server maintainer, at least) from a github repo and a build recipe, I guess. If you agree, we can discuss details like dependencies, static binary vs shared libs etc.

**Comment by frodrigo** [comment]
> Before I embark on a massive fork and language migration.... here's another thought: would it be possible for the osmose-backend script to run a C++ executable on the osmose server? That binary would simply take a string as input and output the sanitized string, and python would just have to do a `subprocess.check_output()` as far as I can see.
> This would allow to truly reuse the existing implementation rather than forking it and having to maintain both in parallel.
> For security reasons I guess I can't provide the binary myself, it would have to be built on the server (or by the server maintainer, at least) from a github repo and a build recipe, I guess. If you agree, we can discuss details like dependencies, static binary vs shared libs etc.

Yes, it is possible to use C++. But rather than running an executable, we can use it as python module with a binding. We already do this for the OSM PBF parser here https://github.com/osm-fr/osmose-backend/tree/master/modules/osm_pbf_parser

**Comment by dfaure-kdab** [comment]
This is *perfect*, I'm writing a boost::python wrapper for a customer this month, so I know how to do that :)

This is definitely less work, and allows us to keep a single code base for the parser.

Yeah, no (other) KDE Frameworks needed, but I'm wondering whether using QtCore (Qt 5) would be ok?

**Comment by dfaure-kdab** [comment]
I'm working on the replacement solution, improving KOpeningHours greatly.

Meanwhile could you disable the current osmose warning? People are making blind changes [1] suggested by osmose, and because they modify opening_hours my import script won't touch it ever again...
[1] https://osmlab.github.io/osm-deep-history/#/way/126800393 for instance

**Comment by dfaure** [comment]
OK the pull request for porting to KOpeningHours (and fixing this issue) is up at https://github.com/osm-fr/osmose-backend/pull/1051

**Comment by frodrigo** [comment]
Parser based on Lark removed.
